### PR TITLE
HDDS-4173.  Implement HDDS Version management using the LayoutVersion…

### PR DESCRIPTION
**Pull request by prashantpogde**
## What changes were proposed in this pull request?

Implement HDDS Version management using the LayoutVersionManager interface.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-4173

## How was this patch tested?

Basic build.

### CODE REVIEW

**Review by avijayanhwx — CHANGES_REQUESTED** [review]
Thanks for working on this @prashantpogde. A few comments inline.

Yesterday I had to reset the upgrade branch since it was in a weird state. Can you rebase your commit on top of the latest branch?

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/storage/ContainerProtocolCalls.java [hunk 1]
```diff
@@ -578,4 +581,35 @@ private static Text getService(DatanodeBlockID blockId) {
     validators.add(validator);
     return validators;
   }
+
+  public static HashMap<DatanodeDetails, GetBlockResponseProto>
+      getBlockFromAllNodes(
```
Comment (linyiqun) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/storage/ContainerProtocolCalls.java:586]: This new introduced method is not used here, what's this method intended used for?

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/ScmSubcommand.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * Base class for admin commands that connect via SCM client.
+ */
+public abstract class ScmSubcommand implements Callable<Void> {
+
+  @CommandLine.Mixin
+  private ScmOption scmOption;
+
+  protected abstract void execute(ScmClient client) throws IOException;
```
Comment (linyiqun) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/ScmSubcommand.java:34]: I see lots of code update focus on ScmSubcommand refactoring, is this related with this JIRA? I don't see there we do some common rolling upgrade logic but seems just a code refactoring.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 3]
```diff
@@ -2244,6 +2248,14 @@ public OmKeyInfo lookupKey(OmKeyArgs args) throws IOException {
     }
   }
 
+
+  @Override
+  public void renameKeys(OmRenameKeys omRenameKeys)
```
Comment (linyiqun) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2253]: RenameKey on Bucket seems also not related with this JIRA if I understand this PR correctly.

**Review by linyiqun — COMMENTED** [review]
@prashantpogde , I take a quick review on this PR. One major thing that I find lots of logic seems not related with this JIRA. Can you have a check for this? There are much change across so many files. Can you create a cleaned PR that will be conveniently reviewed?

File: hadoop-hdds/server-scm/pom.xml [hunk 4]
```diff
@@ -29,6 +29,19 @@ https://maven.apache.org/xsd/maven-4.0.0.xsd">
   <packaging>jar</packaging>
 
   <dependencies>
+
+    <dependency>
```
[the remark this answers is not in the document]
Comment (prashantpogde) [reply hadoop-hdds/server-scm/pom.xml]: done

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/upgrade/SCMLayoutFeatureAspect.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import org.aspectj.lang.annotation.Before;
+import org.aspectj.lang.reflect.MethodSignature;
+
+/**
+ * 'Aspect' for SCM Layout Feature API. All methods annotated with the
+ * specific annotation will have pre-processing done here to check layout
+ * version compatibility.
+ */
+@Aspect
+public class SCMLayoutFeatureAspect {
```
[the remark this answers is not in the document]
Comment (prashantpogde) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/upgrade/SCMLayoutFeatureAspect.java]: done

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/upgrade/SCMLayoutFeatureCatalog.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+/**
+ * Catalog of SCM features.
+ */
+public class SCMLayoutFeatureCatalog {
+
+  /**
+   * List of SCM Features.
+   */
+  public enum SCMLayoutFeature implements LayoutFeature {
```
[the remark this answers is not in the document]
Comment (prashantpogde) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/upgrade/SCMLayoutFeatureCatalog.java]: done

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/upgrade/SCMLayoutFeatureCatalog.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    }
+  }
+
+  /**
+   * This is an example of an "API" that uses a new Layout feature (EC) that is
+   * not yet supported by the current layout version. The following can be
+   * "guarded" by just adding the following annotation, thereby keeping the
+   * method logic and upgrade logic separate.
+   */
+  @SCMLayoutFeatureAPI(SCMLayoutFeature.CREATE_EC)
```
[the remark this answers is not in the document]
Comment (prashantpogde) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/upgrade/SCMLayoutFeatureCatalog.java]: done

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/upgrade/SCMLayoutFeatureCatalog.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+   */
+  public enum SCMLayoutFeature implements LayoutFeature {
+    INITIAL_VERSION(0, "Initial Layout Version"),
+    CREATE_EC(1, ""),
+    NEW_FEATURE(2, "new feature", new NewSCMFeatureUpgradeAction());
+
+
+    private int layoutVersion;
+    private String description;
+    private Optional<SCMUpgradeAction> scmUpgradeAction = Optional.empty();
```
[the remark this answers is not in the document]
Comment (prashantpogde) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/upgrade/SCMLayoutFeatureCatalog.java]: done

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/storage/ContainerProtocolCalls.java (hunk 1 above)
Comment (prashantpogde) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/storage/ContainerProtocolCalls.java]: not sure, why these changes are showed up in this PR. These changes were not made as part of this PR.
I have re-uploaded the changes and it should go away.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/ScmSubcommand.java (hunk 2 above)
Comment (prashantpogde) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/ScmSubcommand.java]: NA. Please see the comment above.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java (hunk 3 above)
Comment (prashantpogde) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java]: NA, like above.

**Comment by prashantpogde** [comment]
Uploaded the changes after addressing comments. Please take a look.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/exceptions/SCMException.java [hunk 9]
```diff
@@ -123,6 +123,7 @@ public ResultCodes getResult() {
     FAILED_TO_INIT_CONTAINER_PLACEMENT_POLICY,
     FAILED_TO_ALLOCATE_ENOUGH_BLOCKS,
     INTERNAL_ERROR,
-    FAILED_TO_INIT_PIPELINE_CHOOSE_POLICY
+    FAILED_TO_INIT_PIPELINE_CHOOSE_POLICY,
+    NOT_SUPPORTED_OPERATION,
```
Comment (linyiqun) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/exceptions/SCMException.java:127]: New SCMException introduced here that broken the unit test:

[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.077 s <<< FAILURE! - in org.apache.hadoop.hdds.scm.exceptions.TestSCMExceptionResultCodes
[ERROR] codeMapping(org.apache.hadoop.hdds.scm.exceptions.TestSCMExceptionResultCodes)  Time elapsed: 0.023 s  <<< FAILURE!
java.lang.AssertionError: expected:<30> but was:<29>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:743)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:555)
	at org.junit.Assert.assertEquals(Assert.java:542)
	at org.apache.hadoop.hdds.scm.exceptions.TestSCMExceptionResultCodes.codeMapping(TestSCMExceptionResultCodes.java:35)

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/upgrade/HDDSLayoutVersionManager.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+import org.apache.hadoop.ozone.upgrade.AbstractLayoutVersionManager;
+import org.apache.hadoop.ozone.upgrade.LayoutVersionManager;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Class to manage layout versions and features for Storage Container Manager
+ * and DataNodes.
+ */
+public final class HDDSLayoutVersionManager extends
```
Comment (linyiqun) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/upgrade/HDDSLayoutVersionManager.java:35]: Can we add the unit test for HDDSLayoutVersionManager like TestOMVersionManager does?

**Review by linyiqun — COMMENTED** [review]
@prashantpogde , PR almost looks great, only two minor comments from me.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/upgrade/NewSCMFeatureUpgradeAction.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+package org.apache.hadoop.hdds.scm.server.upgrade;
+
+import org.apache.hadoop.hdds.scm.server.StorageContainerManager;
+import org.apache.hadoop.hdds.upgrade.HDDSUpgradeAction;
+
+/**
+ * Example SCM Action class to help with understanding.
+ */
+public class NewSCMFeatureUpgradeAction implements
```
Comment (avijayanhwx) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/upgrade/NewSCMFeatureUpgradeAction.java:27]: Unused class?  Can be removed.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/exceptions/SCMException.java (hunk 9 above)
Comment (prashantpogde) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/exceptions/SCMException.java:127]: yup

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/upgrade/HDDSLayoutVersionManager.java (hunk 10 above)
Comment (prashantpogde) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/upgrade/HDDSLayoutVersionManager.java:35]: yup

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/upgrade/NewSCMFeatureUpgradeAction.java (hunk 11 above)
Comment (prashantpogde) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/upgrade/NewSCMFeatureUpgradeAction.java:27]: Yes, left it as an example.  Can we remove it after we are done with other upgrade changes too ?

**Comment by prashantpogde** [comment]
yes, I will remove NOT_SUPPORTED_OPERATION.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/upgrade/HDDSLayoutVersionManager.java (hunk 10 above)
Comment (prashantpogde) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/upgrade/HDDSLayoutVersionManager.java:35]: I will add the test in another PR.

**Review by linyiqun — APPROVED** [review]
Build check all passed now. +1. Feel free to ping me in the next PR, : ).

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/upgrade/NewSCMFeatureUpgradeAction.java (hunk 11 above)
Comment (avijayanhwx) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/upgrade/NewSCMFeatureUpgradeAction.java:27]: Yes, sure.

**Comment by avijayanhwx** [comment]
Thanks for working on this @prashantpogde. LGTM +1
